File: Autonomous_package/src/g2g_obs_avoidance_software.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <iostream>
#include <vector>
#include <string>
#include <bits/stdc++.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;
using namespace std;

class TurtlBot3 : public rclcpp::Node
{
public:
    TurtlBot3()
        : Node("minmal_publsiher"), count_(0)
    {
        subscription_1 = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10, std::bind(&TurtlBot3::topic_callback, this, _1));
        subscription_2 = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10, std::bind(&TurtlBot3::obs_avoid, this, _1));
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
        timer_ = this->create_wall_timer(50ms, std::bind(&TurtlBot3::timer_callback, this));
        TurtlBot3::set_goal();
    }
    vector<float> scan;

private:
    float x_self, y_self, x_goal, y_goal, theta_self;
    long double x, dis_1 = 0, sum = 0, m, n, max, dist;
    void topic_callback(nav_msgs::msg::Odometry msg)
    {
        x_self = msg.pose.pose.position.x;
        y_self = msg.pose.pose.position.y;
        theta_self = quatToEuler(msg.pose.pose.orientation);
    }

    void obs_avoid(const sensor_msgs::msg::LaserScan &msg)
    {
        long double i = 0, k = 0;
        sum = 0;
        for (long double x : msg.ranges) // LiDAR data
        {
            if (x > 2)
            {
                x = 2;
            }
            dis_1 = 1 - x / 2;
            max = (1 - dis_1) * 2;
            scan.push_back(dis_1);
            if (i < 121 || i > 239)
            {
                if (i < 121)
                {
                    k = i;
                }
                else
                {
                    k = i - 359;
                }
                sum -= 3 * dis_1 * k; //calculation of angle of obstacle
            }
            // cout << sum << endl;
            i++;
        }
        scan.clear();
    }

    void set_goal()
    {
        cout << "Enter x coord:" << endl;
        cin >> x_goal;
        cout << "Enter y coord:" << endl;
        cin >> y_goal;
        dist = sqrt(pow(x_goal - x_self, 2) + pow(y_goal - y_self, 2)); // dist of initial posn and final posn
    }

    double quatToEuler(const geometry_msgs::msg::Quaternion &quaternion)
    {
        tf2::Quaternion tf_quat;
        tf2::fromMsg(quaternion, tf_quat);
        double roll, pitch, yaw;
        tf2::Matrix3x3(tf_quat).getRPY(roll, pitch, yaw);   // transformation 
        return yaw;
    }

    void timer_callback()
    {
        double dis, angle, vel = 0, angle_diff = 0, k_lin, obs_lin, obs_ang, k_ang, goal_ang;
        auto message = geometry_msgs::msg::Twist();
        dis = sqrt(pow(x_goal - x_self, 2) + pow(y_goal - y_self, 2));
        if (dis > 0.01)
        {
            angle = atan2(y_goal - y_self, x_goal - x_self) - theta_self;
            if (angle > 3.14)
            {
                angle = angle - 6.28;
            }
            else if (angle < -3.14)
            {
                angle = angle + 6.28; // angle for avoiding infinity
            }
            angle_diff = 0.45 * angle; // angle differnce of the current posn and goal posn
            if (angle_diff > 0.4)
            {
                angle_diff = 0.4 * angle_diff / abs(angle_diff); // limiting goal angular velocity
            }

            vel = (dis / dist);
            if (vel > 0.5)
            {
                vel = 0.5;                                     //limiting goal linear velocity
            }

            obs_lin = (exp(this->dis_1) - 2);
            k_lin = 0.7 - exp(obs_lin);                        // calculator for linear velocity due to obstacle

            goal_ang = 4 / (1 + exp(-500 * (max - 2)));        // calculator for goal angular velocity
            obs_ang = abs(7.389 - exp(goal_ang)) * sum / 600;
            k_ang = 40 * obs_ang + 0.2 * goal_ang * angle_diff;

            message.linear.x = 4 * k_lin * vel;                // adjusting linear velocity
            message.angular.z = k_ang;
            if (abs(message.angular.z) > 0.5)
            {
                message.angular.z = 0.5 * abs(message.angular.z) / message.angular.z; // limiting the resultant angular velocity
            }
            publisher_->publish(message);
            // RCLCPP_INFO(this->get_logger(), "lin_vel : %lf", vel);
            // RCLCPP_INFO(this->get_logger(), "ang_vel : %lf", angle_diff);
        }

        else
        {
            set_goal();
        }
    }
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscription_1;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_2;
    size_t count_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TurtlBot3>());
    rclcpp::shutdown();
    return 0;
}

File: Autonomous package (ros1)/beginner_tutorials/src/g2g_obs_avoidance_software.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist, Pose
from nav_msgs.msg import Odometry
from math import sqrt, atan2, pi, exp
from tf.transformations import euler_from_quaternion

global dis, angle_diff, angle, current_pitch, current_roll, current_yaw, sum, dis_1, max_vals, velocity

def pose_twist_callback(odom_msg):
    global dis, angle_diff, angle, current_pitch, current_roll, current_yaw
    pose = odom_msg.pose.pose
    pose.position.x = round(pose.position.x, 4)
    pose.position.y = round(pose.position.y, 4)
    goal_pose = goal

    dis = sqrt(pow((goal_pose.position.x - pose.position.x), 2) + pow((goal_pose.position.y - pose.position.y), 2))
    current_roll, current_pitch, current_yaw = euler_from_quaternion([pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w])
    angle = atan2(goal_pose.position.y - pose.position.y, goal_pose.position.x - pose.position.x) - current_yaw
    if angle > pi:
        angle -= 2 * pi
    elif angle < -pi:
        angle += 2 * pi  # Angle for avoiding infinity
    angle_diff = min(0.4, 0.45 * abs(angle))*(1 if angle>0 else -1)
    angle_diff = -abs(angle_diff) if angle>2*pi/3 else angle_diff

def avoid_obs(msg):
    global sum, dis_1, max_vals, obs_lin, goal_ang
    sum = 0
    dis_1 = []
    max_vals = []

    for i in range(len(msg.ranges)):
        r = msg.ranges[i]
        r = 1.5 if r > 1.5 else r
        y = 1 - (r / 1.5)
        dis_1.append(y)
        max_vals.append((1 - y) * 1.5)

    for i in range(len(msg.ranges)):
        if i <= 60 or i >= 300:
            sum -= 6 * dis_1[i] * (i if i <= 60 else i - 360)
   
    goal_ang=0
    obs_lin=0
    for index in list(range(0, 61)) + list(range(300, 360)):
        obs_lin = exp(dis_1[index] ) - 1.5
        goal_ang = 4 / (1 + exp(-100 * (max_vals[index] - 1.5)))  # Calculator for goal angular velocity
    

def avoid_obstacle():
    global velocity, vel, angle_diff, k_lin, obs_ang, k_ang, goal_ang
    velocity = dis / orig_dis
    velocity = min(0.5, abs(velocity))*(1 if velocity>0 else -1)  # Limit the velocity to 0.5

    obs_ang = abs(7.389 - exp(goal_ang)) * sum
    k_lin = 0.94 - exp(obs_lin)  # Calculator for linear velocity due to obstacle
    k_ang = obs_ang + 0.3 * goal_ang * angle_diff

    vel.linear.x = 3.5 * k_lin * velocity  # Adjusting linear velocity
    vel.angular.z = k_ang
    vel.angular.z = min(0.4, abs(vel.angular.z)) * (1 if vel.angular.z > 0 else -1)  # Limiting the resultant angular velocity
    pub.publish(vel)
    print(vel.angular.z)
    
if __name__ == '__main__':
    rospy.init_node('fin_g2g', anonymous=True)
    pub = rospy.Publisher('/cmd_vel', Twist, queue_size=10)
    sub_odom = rospy.Subscriber('/odom', Odometry, pose_twist_callback)
    sub_scan = rospy.Subscriber('/scan', LaserScan, avoid_obs)
    global goal
    goal = Pose()
    pose = Pose()
    pose.position.x = round(pose.position.x, 4)
    pose.position.y = round(pose.position.y, 4)
    goal_pose = goal

    goal.position.x = float(input('Enter x coordinate:'))
    goal.position.y = float(input('Enter y coordinate:'))

    vel = Twist()

    rospy.wait_for_message('/odom', Odometry)

    orig_dis = sqrt(pow((goal.position.x - pose.position.x), 2) + pow((goal.position.y - pose.position.y), 2))

    rate = rospy.Rate(20)  # 50ms = 20Hz
    while dis > 0.1 and not rospy.is_shutdown():
        avoid_obstacle()
        rate.sleep()

    vel.linear.x = 0
    vel.angular.z = 0
    pub.publish(vel)
    rospy.spin()
